independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InterceptionProxy.java

```diff
@@ -0,0 +1,29 @@
+package io.quarkus.arc;
+
+/**
+ * A factory for proxies that perform {@link jakarta.interceptor.AroundInvoke}
+ * interception before forwarding the method call to the target instance
+ * ({@code delegate}).
+ * <p>
+ * The class of the target instance is specified by:
+ * <ul>
+ * <li>the type argument to {@code InterceptionProxy}, when declared
+ * as a parameter of a CDI {@linkplain jakarta.enterprise.inject.Produces producer method},
+ * </li>
+ * <li>a parameter of {@code BeanConfigurator.supportsInterception()},
+ * when used in a synthetic bean.</li>
+ * </ul>
+ * This construct only supports {@code @AroundInvoke} interceptors declared on
+ * interceptor classes; other kinds of interception, as well as {@code @AroundInvoke}
+ * interceptors declared on target classes, are not supported.
+ */
+public interface InterceptionProxy<T> {
```

Comment:
Maybe we should do this similarly to how CDI defines `InterceptionFactory` - that it is a *bean* that is provided to you by the container and that you can only use it as param of producer method or in synth beans.
We can then validate and test that it doesn't get injected anywhere else (c-tors, fields, ...) which is a user mistake anyway.

Reply:
I am also thinking that the current approach cannot be used with CDI build compat. extension in any way.
If we managed to turn this into a bean that also carries the bindings source declaration with it (pass in `@BindingsSource` as a fake qualifier perhaps?), we might be able to do that as well, right?

Reply:
> it is a bean that is provided to you by the container and that you can only use it as param of producer method or in synth beans

Agree. I'll rewrite the javadoc, which should be enough, because internally, `InterceptionProxy` already kinda sorta is a bean (a built-in one).

> We can then validate and test that it doesn't get injected anywhere else (c-tors, fields, ...)

I think we already do, see `BuiltinBean.validateInterceptionProxy()`.

> the current approach cannot be used with CDI build compat. extension in any way

Correct.

> If we managed to turn this into a bean that also carries the bindings source declaration with it (pass in `@BindingsSource` as a fake qualifier perhaps?), we might be able to do that as well, right?

Not sure, to be honest. `SyntheticBeanBuilder` has no methods to add injection points or anything like that. And we don't support "injecting" custom parameters into BCEs (yet).

Reply:
> I think we already do, see BuiltinBean.validateInterceptionProxy().

Will do, I probably missed it there.

> Not sure, to be honest. SyntheticBeanBuilder has no methods to add injection points or anything like that. And we don't support "injecting" custom parameters into BCEs (yet).

Hm, right.
We'd probably need to ask users to subclass `SyntheticBeanCreator` or make it extend some new interface so that we knew what the interception proxy requirements are and only then could they use the `Instance` to resolve it. But that sounds a bit wonky, WDYT?

Reply:
Yeah, agree. I would stay away from BCE here.

Reply:
Done, I think.